File: animations/shared/analysis/lpips-analyze.mjs
/**
 * lpips-analyze.mjs
 *
 * Node.js wrapper for LPIPS perceptual similarity analysis.
 * Calls Python script for GPU-accelerated deep learning comparison.
 *
 * Usage:
 *   import { analyzeLPIPS, analyzeLPIPSBatch } from './lpips-analyze.mjs';
 *
 *   // Single pair
 *   const score = await analyzeLPIPS(capturedPath, referencePath);
 *
 *   // Batch (more efficient - loads model once)
 *   const results = await analyzeLPIPSBatch([
 *     { captured: 'a.png', reference: 'ref.png' },
 *     { captured: 'b.png', reference: 'ref.png' }
 *   ]);
 *
 * Output:
 *   LPIPS score (lower = more similar, 0 = identical)
 */

import { spawn } from 'child_process';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const PYTHON_SCRIPT = path.join(__dirname, 'lpips_analyze.py');

/**
 * Check if LPIPS is available (Python + dependencies installed)
 */
export async function isLPIPSAvailable() {
  return new Promise((resolve) => {
    const proc = spawn('python3', ['-c', 'import lpips; import torch']);

    proc.on('close', (code) => {
      resolve(code === 0);
    });

    proc.on('error', () => {
      resolve(false);
    });

    // Timeout after 30 seconds (torch import can be slow)
    setTimeout(() => {
      proc.kill();
      resolve(false);
    }, 30000);
  });
}

/**
 * Analyze LPIPS for a single image pair
 *
 * @param {string} capturedPath - Path to captured image
 * @param {string} referencePath - Path to reference image
 * @param {Object} options - Optional settings
 * @returns {Promise<number|null>} LPIPS score (0-1+) or null on error
 */
export async function analyzeLPIPS(capturedPath, referencePath, options = {}) {
  const { verbose = false, network = 'vgg' } = options;

  return new Promise((resolve, reject) => {
    const args = [
      PYTHON_SCRIPT,
      '--captured', capturedPath,
      '--reference', referencePath,
      '--net', network
    ];

    if (verbose) {
      args.push('--verbose');
    }

    const proc = spawn('python3', args);
    let stdout = '';
    let stderr = '';

    proc.stdout.on('data', (data) => {
      stdout += data.toString();
    });

    proc.stderr.on('data', (data) => {
      stderr += data.toString();
      if (verbose) {
        process.stderr.write(data);
      }
    });

    proc.on('close', (code) => {
      if (code !== 0) {
        console.warn('LPIPS analysis failed:', stderr);
        resolve(null);
        return;
      }

      try {
        const result = JSON.parse(stdout);
        if (result.error) {
          console.warn('LPIPS error:', result.error);
          resolve(null);
        } else {
          resolve(result.lpips);
        }
      } catch (e) {
        console.warn('Failed to parse LPIPS output:', e.message);
        resolve(null);
      }
    });

    proc.on('error', (err) => {
      console.warn('Failed to spawn LPIPS process:', err.message);
      resolve(null);
    });
  });
}

/**
 * Analyze LPIPS for multiple image pairs (batch mode)
 * More efficient than calling analyzeLPIPS multiple times - loads model once.
 *
 * @param {Array<{captured: string, reference: string}>} pairs - Image pairs to analyze
 * @param {Object} options - Optional settings
 * @returns {Promise<Object>} Results with aggregate stats and per-pair scores
 */
export async function analyzeLPIPSBatch(pairs, options = {}) {
  const { verbose = false, network = 'vgg' } = options;

  return new Promise((resolve, reject) => {
    const args = [
      PYTHON_SCRIPT,
      '--batch',
      '--net', network
    ];

    if (verbose) {
      args.push('--verbose');
    }

    const proc = spawn('python3', args);
    let stdout = '';
    let stderr = '';

    // Send pairs as JSON to stdin
    proc.stdin.write(JSON.stringify(pairs));
    proc.stdin.end();

    proc.stdout.on('data', (data) => {
      stdout += data.toString();
    });

    proc.stderr.on('data', (data) => {
      stderr += data.toString();
      if (verbose) {
        process.stderr.write(data);
      }
    });

    proc.on('close', (code) => {
      if (code !== 0) {
        console.warn('LPIPS batch analysis failed:', stderr);
        resolve({
          error: 'LPIPS analysis failed',
          stderr: stderr.slice(-500)
        });
        return;
      }

      try {
        const result = JSON.parse(stdout);
        resolve(result);
      } catch (e) {
        console.warn('Failed to parse LPIPS batch output:', e.message);
        resolve({
          error: 'Failed to parse output',
          stdout: stdout.slice(-500)
        });
      }
    });

    proc.on('error', (err) => {
      console.warn('Failed to spawn LPIPS batch process:', err.message);
      resolve({
        error: err.message
      });
    });
  });
}

/**
 * Analyze LPIPS for animation frames against a reference
 * Convenience wrapper for analyzing multiple captured frames against one reference.
 *
 * @param {string[]} framePaths - Array of captured frame paths
 * @param {string} referencePath - Path to reference image
 * @param {Object} options - Optional settings
 * @returns {Promise<Object>} Results with per-frame scores
 */
export async function analyzeFramesLPIPS(framePaths, referencePath, options = {}) {
  const pairs = framePaths.map(captured => ({
    captured,
    reference: referencePath
  }));

  const result = await analyzeLPIPSBatch(pairs, options);

  if (result.error) {
    return result;
  }

  // Enhance results with frame info
  return {
    ...result,
    frameCount: framePaths.length,
    perFrame: result.results?.map((r, i) => ({
      frame: path.basename(framePaths[i]),
      path: framePaths[i],
      lpips: r.lpips,
      error: r.error
    })) || []
  };
}

// CLI support
const isMainModule = import.meta.url === `file://${process.argv[1]}` ||
                     process.argv[1]?.endsWith('lpips-analyze.mjs');

if (isMainModule) {
  const args = process.argv.slice(2);

  if (args.includes('--help') || args.includes('-h')) {
    console.log(`
lpips-analyze.mjs - Node.js wrapper for LPIPS perceptual similarity

Usage:
  node lpips-analyze.mjs --captured <path> --reference <path>
  node lpips-analyze.mjs --check

Options:
  --captured, -c   Path to captured image
  --reference, -r  Path to reference image
  --check          Check if LPIPS is available
  --verbose, -v    Verbose output
  --help, -h       Show this help

Example:
  node lpips-analyze.mjs -c frame.png -r reference.png
`);
    process.exit(0);
  }

  if (args.includes('--check')) {
    isLPIPSAvailable().then(available => {
      console.log(available ? 'LPIPS is available' : 'LPIPS is NOT available');
      process.exit(available ? 0 : 1);
    });
  } else {
    let captured = null;
    let reference = null;
    let verbose = args.includes('--verbose') || args.includes('-v');

    for (let i = 0; i < args.length; i++) {
      if (args[i] === '--captured' || args[i] === '-c') {
        captured = args[++i];
      } else if (args[i] === '--reference' || args[i] === '-r') {
        reference = args[++i];
      }
    }

    if (!captured || !reference) {
      console.error('Error: --captured and --reference are required');
      process.exit(1);
    }

    analyzeLPIPS(captured, reference, { verbose }).then(score => {
      if (score === null) {
        console.error('LPIPS analysis failed');
        process.exit(1);
      } else {
        console.log(JSON.stringify({ lpips: score }, null, 2));
      }
    });
  }
}

File: animations/shared/analysis/lpips_analyze.py
#!/usr/bin/env python3
"""
lpips_analyze.py

LPIPS (Learned Perceptual Image Patch Similarity) analysis for animation frames.
Uses VGG backbone for perceptual similarity scoring.

Usage:
  # Single pair
  python lpips_analyze.py --captured img1.png --reference img2.png

  # Batch mode (JSON input)
  echo '[{"captured": "a.png", "reference": "b.png"}]' | python lpips_analyze.py --batch

  # Batch mode with file
  python lpips_analyze.py --batch --input pairs.json --output scores.json

Output:
  JSON with LPIPS scores (lower = more similar, 0 = identical)
"""

import argparse
import json
import sys
import os
from pathlib import Path

# Suppress warnings before importing torch
import warnings
warnings.filterwarnings("ignore", category=UserWarning)

import torch
import lpips
from PIL import Image
import numpy as np


def get_device():
    """Detect best available device (CUDA > MPS > CPU)"""
    if torch.cuda.is_available():
        return torch.device("cuda")
    elif hasattr(torch.backends, 'mps') and torch.backends.mps.is_available():
        return torch.device("mps")
    else:
        return torch.device("cpu")


def load_image(path, size=None):
    """Load image and convert to tensor for LPIPS"""
    img = Image.open(path).convert('RGB')

    if size:
        img = img.resize(size, Image.Resampling.LANCZOS)

    # Convert to numpy, normalize to [-1, 1] for LPIPS
    img_np = np.array(img).astype(np.float32) / 255.0
    img_np = img_np * 2.0 - 1.0  # Scale to [-1, 1]

    # Convert to tensor: [H, W, C] -> [1, C, H, W]
    img_tensor = torch.from_numpy(img_np).permute(2, 0, 1).unsqueeze(0)

    return img_tensor


def compute_lpips(model, captured_path, reference_path, device):
    """Compute LPIPS distance between two images"""
    try:
        # Load images
        captured = load_image(captured_path)
        reference = load_image(reference_path)

        # Resize if dimensions don't match
        if captured.shape != reference.shape:
            # Resize captured to match reference
            ref_size = (reference.shape[3], reference.shape[2])  # (W, H)
            captured = load_image(captured_path, size=ref_size)

        # Move to device
        captured = captured.to(device)
        reference = reference.to(device)

        # Compute LPIPS
        with torch.no_grad():
            distance = model(captured, reference)

        return float(distance.cpu().numpy().flatten()[0])

    except Exception as e:
        return {"error": str(e)}


def process_batch(pairs, model, device, verbose=False):
    """Process a batch of image pairs"""
    results = []

    for i, pair in enumerate(pairs):
        captured = pair.get("captured")
        reference = pair.get("reference")

        if not captured or not reference:
            results.append({
                "captured": captured,
                "reference": reference,
                "error": "Missing captured or reference path"
            })
            continue

        if not os.path.exists(captured):
            results.append({
                "captured": captured,
                "reference": reference,
                "error": f"Captured file not found: {captured}"
            })
            continue

        if not os.path.exists(reference):
            results.append({
                "captured": captured,
                "reference": reference,
                "error": f"Reference file not found: {reference}"
            })
            continue

        if verbose:
            print(f"  [{i+1}/{len(pairs)}] {os.path.basename(captured)}", file=sys.stderr)

        score = compute_lpips(model, captured, reference, device)

        if isinstance(score, dict) and "error" in score:
            results.append({
                "captured": captured,
                "reference": reference,
                "error": score["error"]
            })
        else:
            results.append({
                "captured": captured,
                "reference": reference,
                "lpips": score
            })

    return results


def main():
    parser = argparse.ArgumentParser(description="LPIPS perceptual similarity analysis")
    parser.add_argument("--captured", "-c", help="Path to captured image")
    parser.add_argument("--reference", "-r", help="Path to reference image")
    parser.add_argument("--batch", "-b", action="store_true", help="Batch mode (read JSON from stdin or --input)")
    parser.add_argument("--input", "-i", help="Input JSON file for batch mode")
    parser.add_argument("--output", "-o", help="Output JSON file (default: stdout)")
    parser.add_argument("--net", default="vgg", choices=["vgg", "alex", "squeeze"], help="Network backbone")
    parser.add_argument("--verbose", "-v", action="store_true", help="Verbose output")

    args = parser.parse_args()

    # Detect device
    device = get_device()
    if args.verbose:
        print(f"Using device: {device}", file=sys.stderr)

    # Load model
    if args.verbose:
        print(f"Loading LPIPS model ({args.net})...", file=sys.stderr)

    model = lpips.LPIPS(net=args.net, verbose=False).to(device)
    model.eval()

    if args.verbose:
        print("Model loaded.", file=sys.stderr)

    # Process based on mode
    if args.batch:
        # Batch mode
        if args.input:
            with open(args.input, 'r') as f:
                pairs = json.load(f)
        else:
            pairs = json.load(sys.stdin)

        results = process_batch(pairs, model, device, args.verbose)

        # Compute aggregate stats
        valid_scores = [r["lpips"] for r in results if "lpips" in r]
        aggregate = {
            "count": len(results),
            "valid": len(valid_scores),
            "errors": len(results) - len(valid_scores)
        }

        if valid_scores:
            aggregate["mean"] = sum(valid_scores) / len(valid_scores)
            aggregate["min"] = min(valid_scores)
            aggregate["max"] = max(valid_scores)

        output = {
            "device": str(device),
            "network": args.net,
            "aggregate": aggregate,
            "results": results
        }

    else:
        # Single pair mode
        if not args.captured or not args.reference:
            parser.error("--captured and --reference required in single mode")

        score = compute_lpips(model, args.captured, args.reference, device)

        if isinstance(score, dict) and "error" in score:
            output = {
                "device": str(device),
                "network": args.net,
                "captured": args.captured,
                "reference": args.reference,
                "error": score["error"]
            }
        else:
            output = {
                "device": str(device),
                "network": args.net,
                "captured": args.captured,
                "reference": args.reference,
                "lpips": score
            }

    # Output
    output_json = json.dumps(output, indent=2)

    if args.output:
        with open(args.output, 'w') as f:
            f.write(output_json)
        if args.verbose:
            print(f"Results written to: {args.output}", file=sys.stderr)
    else:
        print(output_json)


if __name__ == "__main__":
    main()
